File: packages/python/elizaos/features/basic_capabilities/providers/capabilities.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_capabilities(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    model_types = ["TEXT_LARGE", "TEXT_SMALL", "TEXT_EMBEDDING", "IMAGE", "AUDIO"]
    available_models = [mt for mt in model_types if runtime.has_model(mt)]
    service_names = [s.name for s in runtime.services if hasattr(s, "name")]

    features: list[str] = []
    if runtime.get_setting("ENABLE_VOICE"):
        features.append("voice")
    if runtime.get_setting("ENABLE_VISION"):
        features.append("vision")
    if runtime.get_setting("ENABLE_MEMORY"):
        features.append("long_term_memory")

    capabilities: dict[str, list[str] | bool] = {
        "models": available_models,
        "services": service_names,
        "features": features,
    }

    text_parts: list[str] = ["# Agent Capabilities"]

    if available_models:
        text_parts.append(f"Models: {', '.join(available_models)}")

    if service_names:
        text_parts.append(f"Services: {', '.join(service_names)}")

    if features:
        text_parts.append(f"Features: {', '.join(features)}")

    return ProviderResult(
        text="\n".join(text_parts),
        values={
            "modelCount": len(available_models),
            "serviceCount": len(service_names),
            "hasVoice": "voice" in features,
            "hasVision": "vision" in features,
        },
        data=capabilities,
    )


capabilities_provider = Provider(
    name="CAPABILITIES",
    description="Agent capabilities including models, services, and features",
    get=get_capabilities,
    dynamic=False,
)

File: packages/rust/src/features/basic_capabilities/bootstrap/providers/capabilities.rs
//! CAPABILITIES provider implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::PluginResult;
use crate::generated::spec_helpers::require_provider_spec;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ModelType, ProviderResult, State};

use super::Provider;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ProviderDoc> =
    Lazy::new(|| require_provider_spec("CAPABILITIES"));

/// Provider for agent capabilities.
pub struct CapabilitiesProvider;

#[async_trait]
impl Provider for CapabilitiesProvider {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    fn is_dynamic(&self) -> bool {
        SPEC.dynamic.unwrap_or(false)
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        _message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        // Check available models
        let model_types = [
            (ModelType::TextLarge, "TEXT_LARGE"),
            (ModelType::TextSmall, "TEXT_SMALL"),
            (ModelType::TextEmbedding, "TEXT_EMBEDDING"),
            (ModelType::Image, "IMAGE"),
            (ModelType::AudioTranscription, "AUDIO"),
        ];

        let available_models: Vec<&str> = model_types
            .iter()
            .filter(|(model_type, _)| runtime.has_model(*model_type))
            .map(|(_, name)| *name)
            .collect();

        // Check features based on settings
        let mut features: Vec<&str> = Vec::new();
        if runtime.get_setting("ENABLE_VOICE").is_some() {
            features.push("voice");
        }
        if runtime.get_setting("ENABLE_VISION").is_some() {
            features.push("vision");
        }
        if runtime.get_setting("ENABLE_MEMORY").is_some() {
            features.push("long_term_memory");
        }

        let mut text_parts: Vec<String> = vec!["# Agent Capabilities".to_string()];

        if !available_models.is_empty() {
            text_parts.push(format!("Models: {}", available_models.join(", ")));
        }

        if !features.is_empty() {
            text_parts.push(format!("Features: {}", features.join(", ")));
        }

        Ok(ProviderResult::new(text_parts.join("\n"))
            .with_value("modelCount", available_models.len() as i64)
            .with_value("hasVoice", features.contains(&"voice"))
            .with_value("hasVision", features.contains(&"vision"))
            .with_data(
                "models",
                serde_json::to_value(&available_models).unwrap_or_default(),
            )
            .with_data(
                "features",
                serde_json::to_value(&features).unwrap_or_default(),
            ))
    }
}
